optimize tidy check on `src/tools/tidy/src/issues.txt`

onur-ozkan opened the pull request:
This change applies to the following:

- Handles `is_sorted` in the first iteration without needing a second.
- Fixes line sorting on `--bless`.
- Reads `issues.txt` as str rather than making it part of the source code.

Fixes #123002

workingjubilee commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -186,29 +196,19 @@ pub fn check(root_path: &Path, bless: bool, bad: &mut bool) {
     // if there are any file names remaining, they were moved on the fs.
     // our data must remain up to date, so it must be removed from issues.txt
     // do this automatically on bless, otherwise issue a tidy error
-    if bless && !remaining_issue_names.is_empty() {
-        let issues_txt_header = r#"
-/*
-============================================================
-    ⚠️⚠️⚠️NOTHING SHOULD EVER BE ADDED TO THIS LIST⚠️⚠️⚠️
-============================================================
-*/
-[
-"#;
```
I genuinely do not think we should remove this.

onur-ozkan replied:
I understand the concern but I'm not sure how this will help. If someone adds a line to the end of this file, reviewers won't be able to see this header anyway.

How about hardcoding the current length and checking if it's not equal to that? This way we will continuously be aware of it without relying on reviewers' mistakes.

workingjubilee replied:
It is not for informing a reviewer.

`const ISSUES_ENTRY_LIMIT` is already checked.

onur-ozkan replied:
> It is not for informing a reviewer.

It has to. When someone accidentally adds a line, reviewer should be aware of this information. 

> const ISSUES_ENTRY_LIMIT is already checked.

What I mean is this:

```diff
diff --git a/src/tools/tidy/src/ui_tests.rs b/src/tools/tidy/src/ui_tests.rs
index 5b3b948d4a1..86e72823036 100644
--- a/src/tools/tidy/src/ui_tests.rs
+++ b/src/tools/tidy/src/ui_tests.rs
@@ -100,6 +100,9 @@ fn check_entries(tests_path: &Path, bad: &mut bool) {
 }

 pub fn check(root_path: &Path, bless: bool, bad: &mut bool) {
+    // This can be decreased but can never be increased.
+    const EXPECTED_LINE_COUNT: usize = 4372;
+
     let path = &root_path.join("tests");
     check_entries(&path, bad);

@@ -120,6 +123,8 @@ pub fn check(root_path: &Path, bless: bool, bad: &mut bool) {
         })
         .collect();

+    assert_eq!(EXPECTED_LINE_COUNT, allowed_issue_names.len(), "If you have added new lines to issues.txt, please revert them; otherwise, decrease the EXPECTED_LINE_COUNT.");
+
     if !is_sorted && !bless {
         tidy_error!(
             bad,
```

workingjubilee replied:
> It has to. When someone accidentally adds a line, reviewer should be aware of this information.

The vast majority of the information in the Rust repository is addressed directly at the contributor, and only incidentally useful to a reviewer, so I disagree with this "has to" assertion.

workingjubilee replied:
Anyways, the reason there's nothing exactingly doing `assert_eq!` is because there was a request in the original PR to make this list `--bless`able, to reduce annoyance. So I implemented that. Making sure this number `--bless`es only up and not down would require a lot of fiddly logic, I would think? And it would also invite merge conflicts on the value a lot.

bors commented:
:umbrella: The latest upstream changes (presumably #123429) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Mark-Simulacrum commented:
It seems like we ought to be able to move to using include_str!() and skip a fixed header before iterating through the lines, without making any other, unrelated changes to the behavior (e.g., adding an assertion about a specific number).

onur-ozkan committed:
optimize tidy check on `src/tools/tidy/src/issues.txt`

This change applies to the following:

- Handles `is_sorted` in the first iteration without needing a second.
- Fixes line sorting on `--bless`.
- Reads `issues.txt` as str rather than making it part of the source code.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan commented:
@rustbot ready

Mark-Simulacrum commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -100,23 +100,39 @@ fn check_entries(tests_path: &Path, bad: &mut bool) {
+    let mut is_sorted = true;
+    let allowed_issue_names: BTreeSet<_> = include_str!("issues.txt")
+        .lines()
+        .skip(issues_txt_header.lines().count()) // skip the header lines
```
This should probably be a strip_prefix(issues_txt_header), so that we assert that the contents of that block have not changed.

Mark-Simulacrum commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -100,23 +100,39 @@ fn check_entries(tests_path: &Path, bad: &mut bool) {
+                is_sorted = false;
+            }
+
+            prev_line = line.clone();
```
FWIW, there's some needless copying going on here AFAICT. `BTreeSet<&'static str>` should be fine (and should be what you get from iterating lines).

Mark-Simulacrum reviewed (commented):
r=me with nits fixed

onur-ozkan commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 49c10e4fec7e365eb700eb2115351422d54719a0 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 49c10e4fec7e365eb700eb2115351422d54719a0 -->
<!-- homu: {"type":"Approved","sha":"49c10e4fec7e365eb700eb2115351422d54719a0","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- 
https://github.com/rust-lang/rust/pull/123544#issuecomment-2041047909

onur-ozkan commented:
@bors r=Mark-Simulacrum (corrected the usage of `/` in Windows paths.)

bors commented:
:pushpin: Commit b0f2e60a672f6be6d16d420fc679beb9b3212f3f has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum b0f2e60a672f6be6d16d420fc679beb9b3212f3f -->
<!-- homu: {"type":"Approved","sha":"b0f2e60a672f6be6d16d420fc679beb9b3212f3f","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b0f2e60a672f6be6d16d420fc679beb9b3212f3f with merge 773fb88e13b06174d0ba48191731ea43ab6ac668...
<!-- homu: {"type":"BuildStarted","head_sha":"b0f2e60a672f6be6d16d420fc679beb9b3212f3f","merge_sha":"773fb88e13b06174d0ba48191731ea43ab6ac668"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8582033272/job/23520960632)
Approved by: Mark-Simulacrum
Pushing 773fb88e13b06174d0ba48191731ea43ab6ac668 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8582033272/job/23520960632"},"merge_sha":"773fb88e13b06174d0ba48191731ea43ab6ac668"} -->

rust-timer commented:
Finished benchmarking commit (773fb88e13b06174d0ba48191731ea43ab6ac668): [comparison URL](https://perf.rust-lang.org/compare.html?start=83d0a940c65e9c276308c81107b9b21cf399cbc6&end=773fb88e13b06174d0ba48191731ea43ab6ac668&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=83d0a940c65e9c276308c81107b9b21cf399cbc6&end=773fb88e13b06174d0ba48191731ea43ab6ac668&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=83d0a940c65e9c276308c81107b9b21cf399cbc6&end=773fb88e13b06174d0ba48191731ea43ab6ac668&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=83d0a940c65e9c276308c81107b9b21cf399cbc6&end=773fb88e13b06174d0ba48191731ea43ab6ac668&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [2.2%, 2.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=83d0a940c65e9c276308c81107b9b21cf399cbc6&end=773fb88e13b06174d0ba48191731ea43ab6ac668&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 666.761s -> 668.041s (0.19%)
**Artifact size**: 318.27 MiB -> 318.25 MiB (-0.01%)
